Use the full display name of projects in notifications

yrodiere opened the pull request:
Fixes #31.

Use the full display name of projects in notifications instead of the "relative" name.

I.e. a job named "TestJob" in a folder named "Folder" will be displayed as "Folder » TestJob" instead of just "TestJob".

More importantly, a job for branch "production" in a multi-branch pipeline project named "TestJob" will be displayed as "TestJob » production" instead of just "production".

- [x] Make sure you are opening from a **topic/feature/bugfix branch** (right side) and not your main branch!
- [x] Ensure that the pull request title represents the desired changelog entry
- [x] Please describe what you did
- [x] Link to relevant issues in GitHub or Jira
- [ ] Link to relevant pull requests, esp. upstream and downstream changes
- [x] Ensure you have provided tests - that demonstrates feature works or fixes the issue

<!--
Put an `x` into the [ ] to show you have filled the information.
The template comes from https://github.com/jenkinsci/.github/blob/master/.github/pull_request_template.md 
You can override it by creating .github/pull_request_template.md  in your own repository 
-->

yrodiere commented:
Hey @butchyyyy ; any chance to get this merged and released? Do I need to change anything?

butchyyyy commented:
Hi, thank your for taking the time to contribute to the Jenkin's zulip-plugin. I'm sorry for the super late reply - as you probably figured out - I'm not actively maintaining this plugin anymore.

Couple comments regarding the change:
- Instead of using `getFullDisplayName()`, you could loop through build parents. The advantage being you would get separate url to branch project and the whole multi branch pipeline (similar to what we have now where there is a separate url for build and for branch project).
- To be consistent, you should probably also use the full project's name as topic name in case the notifier is not posting into a fixed one.

I assume that your use case is that your team maintains multiple multi branch pipelines and post them all into same stream + topic. I have a slight concern that people that have dedicated stream + topic for each multi branch pipeline will get additional verbosity that they don't need. Can you give introducing a configuration options for this a thought? It would default to what we have now with the option to display the full name instead.

yrodiere committed:
Refactor link generation methods in ZulipNotifier

So that we can generate multiple links mor easily in the next commits.

yrodiere committed:
Add configuration options to use full job paths in messages/topics

yrodiere commented:
@butchyyyy Thanks for the feedback. Here is an updated version adjusted according to your comments. I also made sure to update `ZulipSendStep`, which I had missed originally.

Does this look good? The tests seem to pass, but I'm not sure what I did about the configuration options ir right: I've never added a configuration option to a Jenkins plugin before.

butchyyyy commented on src/main/java/jenkins/plugins/zulip/ZulipUtil.java:
```diff
@@ -73,4 +76,68 @@ public static String expandVariables(@Nonnull Run<?, ?> run, @Nonnull TaskListen
+     * @param displayLinks Whether to display links to the job and (if relevant) its parent items.
+     * @return A string representing the job.
+     */
+    public static String displayJob(Job<?, ?> job, DescriptorImpl globalConfig, boolean fullJobPath, boolean displayLinks) {
```
nit: Can you add tests for the new helper methods?

butchyyyy commented on src/main/resources/jenkins/plugins/zulip/ZulipNotifier/global.jelly:
```diff
@@ -29,6 +29,15 @@
+    <f:entry title="Use full job path in notification message" help="/plugin/zulip/help-globalConfig-fullJobPathInMessage.html">
+        <f:checkbox name="fullJobPathInMessage" checked="${descriptor.isFullJobPathInMessage()}" />
+    </f:entry>
+    <f:entry title="Enable Smart Notification" help="/plugin/zulip/help-globalConfig-smartNotify.html">
```
Duplicates the smart notification settings.

yrodiere replied:
Sorry, that was a copy/paste error. I'll remove it.

butchyyyy reviewed (commented):
This is awesome! Couple more comments (one might be the reason of integration test failure).

I'll work on my local set up to give this a quick test on my end and to be able to release this eventually. Please bear with me if this takes couple more days as I did not do anything Jenkins or Java related for over a year.

yrodiere commented:
I addressed your comments. Thanks for the review, and thanks for your work on the release :)

If you haven't worked on this project or on Java projects for a while, this piece of information may be useful: this project apparently only compiles on Java 8. That shouldn't impact end-users, but if you're reinstalling everything, make sure you install JDK 8, not 11 or 17.

butchyyyy commented on src/main/java/jenkins/plugins/zulip/DescriptorImpl.java:
```diff
@@ -94,6 +96,22 @@ public void setTopic(String topic) {
         this.topic = topic;
     }
 
+    public boolean isFullJobPathAsDefaultTopic() {
```
You also need to add the new fields to the `configure` method:

```java
fullJobPathAsDefaultTopic = Boolean.TRUE.equals(json.get("fullJobPathAsDefaultTopic"));
fullJobPathInMessage = Boolean.TRUE.equals(json.get("fullJobPathInMessage"));
```

butchyyyy commented on src/main/java/jenkins/plugins/zulip/ZulipUtil.java:
```diff
@@ -73,4 +76,69 @@ public static String expandVariables(@Nonnull Run<?, ?> run, @Nonnull TaskListen
+                                      boolean fullPath, boolean displayLinks) {
+        // The only common interface between Item and ItemGroup is ModelObject, which doesn't define getParent,
+        // so we need to resort to instanceof + cast to crawl up the item tree.
+        if (fullPath && object instanceof Item) {
```
Traversing all the way to the top will also print out the Jenkins instance itself which is something we most likely want to avoid.
![image](https://user-images.githubusercontent.com/17402486/135939601-41ccaf96-8614-45ff-a64a-88660802ef86.png)

One possibility is to add something like this to the beginning of the method:

```java
if (object instanceof jenkins.model.Jenkins) {
    return;
}
```

butchyyyy reviewed (commented):
Got around to test this and found couple problems, once they are addressed this will be ready to merge and release.

yrodiere committed:
Fix "fullJobPath*" configuration deserialization in DescriptorImpl

yrodiere committed:
Ignore Jenkins when displaying full job paths

yrodiere commented:
Thanks, and sorry for not noticing that, I guess I should have spun up a Jenkins instance instead of trusting my tests blindly :/

I added two commits to fix the problems you reported.

butchyyyy commented:
Release 1.3.0 has been published to Jenkins artifactory and the new version should be available in update centre in under 24hours.
Thanks for being patient with this dragging for so long (was super busy at work GAing new product).

Let me know if you run into any issue.

yrodiere commented:
I installed the update and it work. Thank you very much for the reviews and release!
